lib/inc/internal/sources/cpuid_source.hpp

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <string>
+
+namespace whereami { namespace sources {
+
+    /**
+     * Register values returned by CPUID
+     */
+    struct cpuid_registers
+    {
+        /**
+         * eax value
+         */
+        unsigned int eax;
+        /**
+         * ebx value
+         */
+        unsigned int ebx;
+        /**
+         * ecx value
+         */
+        unsigned int ecx;
+        /**
+         * edx value
+         */
+        unsigned int edx;
+    };
+
+    /**
+     * Base cpuid data source
+     */
+    class cpuid_base
+    {
+    public:
+        /**
+         * Call the cpuid instruction
+         * @param leaf The value to pass into CPUID via eax; Determines the type of information returned
+         * @return Register object with results of the CPUID function
+         */
+        virtual cpuid_registers read_cpuid(unsigned int leaf) const;
+        /**
+         * Determine whether CPUID reports that this system is running on a hypervisor (Calls CPUID with eax = 1)
+         * @return whether CPUID reports a hypervisor
+         */
+        bool has_hypervisor() const;
+        /**
+         * Retrieve the vendor ID (Calls CPUID with eax = INFO_LEAF)
+         * @return the vendor ID string
+         */
+        std::string vendor() const;
+        /**
+         * Most hypervisors store vendor information in this leaf
+         */
+        constexpr static const unsigned int INFO_LEAF = 0x40000000;
```

Comment:
Did something require this to be flagged as `constexpr`? Usually just being a `static const` should be enough to get the compiler to inline the value, unless it's involved in other `constexpr` functions